libraries/tools/kotlin-gradle-plugin-core/src/main/kotlin/org/jetbrains/kotlin/gradle/tasks/Tasks.kt

```diff
@@ -163,16 +217,341 @@ public open class KotlinCompile() : AbstractKotlinCompile<K2JVMCompilerArguments
         logger.kotlinDebug("args.moduleName = ${args.moduleName}")
     }
 
+    override fun callCompiler(args: K2JVMCompilerArguments, sources: List<File>, isIncrementalRequested: Boolean, modified: List<File>, removed: List<File>, cachesBaseDir: File) {
+
+        if (experimentalIncremental) {
+            // TODO: consider other ways to pass incremental flag to compiler/builder
+            System.setProperty("kotlin.incremental.compilation", "true")
+            // TODO: experimental should be removed as soon as it becomes standard
+            System.setProperty("kotlin.incremental.compilation.experimental", "true")
+        }
+
+        val targetType = "java-production"
+        val moduleName = args.moduleName
+        val targets = listOf(TargetId(moduleName, targetType))
+        val outputDir = File(args.destination)
+        val caches = hashMapOf<TargetId, GradleIncrementalCacheImpl>()
+        val lookupStorage = LookupStorage(File(cachesBaseDir, "lookups"))
+        val lookupTracker = LookupTrackerImpl(LookupTracker.DO_NOTHING)
+        var currentRemoved = removed
+        val allGeneratedFiles = hashSetOf<GeneratedFile<TargetId>>()
+        val compiledSourcesSet = hashSetOf<File>()
+
+        fun getOrCreateIncrementalCache(target: TargetId): GradleIncrementalCacheImpl {
+            val cacheDir = File(cachesBaseDir, "increCache.${target.name}")
+            cacheDir.mkdirs()
+            return GradleIncrementalCacheImpl(targetDataRoot = cacheDir, targetOutputDir = outputDir, target = target)
+        }
+
+        fun getIncrementalCache(it: TargetId) = caches.getOrPut(it, { getOrCreateIncrementalCache(it) })
+
+        fun PsiClass.findLookupSymbols(): Iterable<LookupSymbol> {
+            val fqn = qualifiedName.orEmpty()
+            return listOf(LookupSymbol(name.orEmpty(), if (fqn == name) "" else fqn.removeSuffix("." + name!!))) +
+                    methods.map { LookupSymbol(it.name, fqn) } +
+                    fields.map { LookupSymbol(it.name.orEmpty(), fqn) } +
+                    innerClasses.flatMap { it.findLookupSymbols() }
+        }
+
+        fun Iterable<LookupSymbol>.files(filesFilter: (File) -> Boolean = { true }, logAction: (LookupSymbol, Iterable<File>) -> Unit = { l,fs -> }): Iterable<File> =
```

Comment:
1. I never paid attention to that part, but I'll have a look.
2,3. I strongly believe that almost any code should be written as close to library design principles as possible without hurting performance, readability, simplicity (to certain extent), and may be few other constraints. Because if it is written properly it could be reused. Hence the generalizations there that going beyond actual usage at the moment.